File: app/src/main/java/org/descartae/android/view/fragments/facility/FeedbackDialog.java
package org.descartae.android.view.fragments.facility;

import android.app.AlertDialog;
import android.app.Dialog;
import android.os.Bundle;
import android.support.design.widget.Snackbar;
import android.support.v4.app.DialogFragment;
import android.view.ContextThemeWrapper;
import android.view.View;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.TextView;

import org.descartae.android.AddFeedbackMutation;
import org.descartae.android.DescartaeApp;
import org.descartae.android.R;
import org.descartae.android.presenter.feedback.FeedbackPresenter;
import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;
import org.greenrobot.eventbus.ThreadMode;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

/**
 * Created by lucasmontano on 09/12/2017.
 */
public class FeedbackDialog extends DialogFragment {

    @Inject FeedbackPresenter presenter;

    @Inject EventBus eventBus;

    @BindView(R.id.title)
    public TextView mTitle;

    @BindView(R.id.subtitle)
    public TextView mSubTitle;

    @BindView(R.id.message)
    public EditText mFeedback;

    @BindView(R.id.action_cancel)
    public View mActionCancel;

    @BindView(R.id.action_send)
    public View mActionSend;

    @BindView(R.id.action_ok)
    public View mActionOk;

    private String facilityID;
    private AlertDialog.Builder mBuilder;

    public static FeedbackDialog newInstance(String facilityID) {
        FeedbackDialog frag = new FeedbackDialog();
        Bundle args = new Bundle();
        args.putString("facilityID", facilityID);
        frag.setArguments(args);
        return frag;
    }

    public void onActivityCreated(Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);

        /*
         * Init Dagger
         */
        DescartaeApp.getInstance(getActivity())
                .getAppComponent()
                .inject(this);

        presenter.setFacilityId(facilityID);
    }

    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {

        LinearLayout viewInflated = (LinearLayout) getActivity().getLayoutInflater().inflate(R.layout.dialog_feedback, null);

        ButterKnife.bind(this, viewInflated);

        facilityID = getArguments().getString("facilityID");

        if (facilityID == null) {
            mTitle.setText(R.string.feedback_title);
            mSubTitle.setText(R.string.feedback_desc);
        } else {
            mTitle.setText(R.string.feedback_facility_title);
            mSubTitle.setText(R.string.feedback_facility_desc);
        }

        mBuilder = new AlertDialog.Builder(new ContextThemeWrapper(getActivity(), R.style.Theme_AppCompat_Light));
        mBuilder.setView(viewInflated);

        return mBuilder.create();
    }

    @Override
    public void onStart() {
        super.onStart();
        eventBus.register(this);
    }

    @Override
    public void onStop() {
        super.onStop();
        eventBus.unregister(this);
    }

    @OnClick(R.id.action_send)
    public void onSend() {

        String feedback = mFeedback.getText().toString();

        if (feedback == null || feedback.length() <= 0) {
            Snackbar.make(mFeedback, R.string.feedback_no_message_error, Snackbar.LENGTH_SHORT).show();
            return;
        }

        presenter.sendFeedback(feedback);
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    public void onFeedbackResult(AddFeedbackMutation.Data data) {
        if (data.addFeedback().success()) {
            mActionCancel.setVisibility(View.GONE);
            mActionSend.setVisibility(View.GONE);
            mActionOk.setVisibility(View.VISIBLE);

            mFeedback.setVisibility(View.GONE);

            if (facilityID == null) {
                mTitle.setText(R.string.feedback_title_success);
                mSubTitle.setText(R.string.feedback_desc_success);
            } else {
                mTitle.setText(R.string.feedback_facility_title_success);
                mSubTitle.setText(R.string.feedback_facility_desc_success);
            }
        } else {
            Snackbar.make(mFeedback, R.string.feedback_error, Snackbar.LENGTH_SHORT).show();
        }
    }

    @OnClick({R.id.action_cancel, R.id.action_ok})
    public void onClose() {
        dismiss();
    }
}

File: app/src/main/java/org/descartae/android/view/fragments/facility/FeedbackDialog.kt
package org.descartae.android.view.fragments.facility

import android.os.Bundle
import android.support.design.widget.Snackbar
import android.support.v4.app.DialogFragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import kotlinx.android.synthetic.main.dialog_feedback.*
import org.descartae.android.AddFeedbackMutation
import org.descartae.android.DescartaeApp
import org.descartae.android.R
import org.descartae.android.networking.apollo.errors.GeneralError
import org.descartae.android.presenter.feedback.FeedbackPresenter
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode
import javax.inject.Inject

class FeedbackDialog : DialogFragment() {

    @Inject lateinit var presenter: FeedbackPresenter

    @Inject lateinit var eventBus: EventBus

    private var facilityID: String? = null

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        (activity!!.applicationContext as DescartaeApp).component.inject(this)

        facilityID = arguments?.getString("facilityID")

        presenter.setFacilityId(facilityID)

        if (facilityID == null) {
            textView_title.text = getString(R.string.feedback_title)
            textView_subtitle.text = getString(R.string.feedback_desc)
        } else {
            textView_title.text = getString(R.string.feedback_facility_title)
            textView_subtitle.text = getString(R.string.feedback_facility_desc)
        }

        button_cancel.setOnClickListener { dismiss() }
        button_ok.setOnClickListener { dismiss() }
        button_send.setOnClickListener {

            val feedback: String = editText_message.text.toString()

            if (feedback.isEmpty())
                Snackbar.make(editText_message, R.string.feedback_no_message_error, Snackbar.LENGTH_SHORT).show()
            else presenter.sendFeedback(feedback)
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.dialog_feedback, container, false)
    }

    override fun onStart() {
        super.onStart()
        eventBus.register(this)
    }

    override fun onStop() {
        super.onStop()
        eventBus.unregister(this)
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onFeedbackResult(data: AddFeedbackMutation.Data) {
        if (data.addFeedback().success()) {
            button_cancel.visibility = View.GONE
            button_send.visibility = View.GONE
            button_ok.visibility = View.VISIBLE

            editText_message.visibility = View.GONE

            if (facilityID == null) {
                textView_title.setText(R.string.feedback_title_success)
                textView_subtitle.setText(R.string.feedback_desc_success)
            } else {
                textView_title.setText(R.string.feedback_facility_title_success)
                textView_subtitle.setText(R.string.feedback_facility_desc_success)
            }
        } else {
            Snackbar.make(editText_message, R.string.feedback_error, Snackbar.LENGTH_SHORT).show()
        }
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onError(error: GeneralError) {
        dismiss()
    }

    companion object {

        fun newInstance(facilityID: String?): FeedbackDialog {
            val frag = FeedbackDialog()
            val args = Bundle()
            args.putString("facilityID", facilityID)
            frag.arguments = args
            return frag
        }
    }
}
